pkg/process/util/api/weighted_queue.go

```diff
@@ -0,0 +1,208 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package api
+
+import (
+	"container/list"
+	"sync"
+)
+
+// WeightedItem is an item that has a type and weight that can be added to a WeightedQueue
+type WeightedItem interface {
+	// Weight returns the weight of this item
+	Weight() int64
+	// Type returns the type of this item
+	Type() string
+}
+
+// WeightedQueue is a queue of WeightedItems.
+//
+// The queue is configured with a maximum size (the maximum number of elements allowed in the queue) as well as
+// a maximum weight.  If adding an item to the queue would violate either the max weight or max size, earlier items
+// are purged from the queue until there is room for the newest item.
+//
+// Items added to the queue have a weight and type.  When purging existing items to make room for new, items of the same
+// type being added will be removed first before moving on to other types.
+type WeightedQueue struct {
+	// dataAvailable is a channel used for communication between callers invoking the Add method and callers
+	// blocked on Poll.  If a caller calls Poll and the queue is empty, the call will block waiting for a send
+	// on dataAvailable.  When Add is invoked, it will perform a non-blocking send on dataAvailable to notify the caller
+	// blocked on Poll
+	dataAvailable chan struct{}
+	queue         *list.List
+	mu            sync.Mutex
+
+	maxSize       int
+	maxWeight     int64
+	currentWeight int64
+}
+
+// NewWeightedQueue returns a new WeightedQueue with the given maximum size & weight
+func NewWeightedQueue(maxSize int, maxWeight int64) *WeightedQueue {
+	return &WeightedQueue{
+		dataAvailable: make(chan struct{}),
+		queue:         list.New(),
+		maxSize:       maxSize,
+		maxWeight:     maxWeight,
+		currentWeight: 0,
+	}
+}
+
+// Len returns the number of items in the queue
+func (q *WeightedQueue) Len() int {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.queue.Len()
+}
+
+// Weight returns the current weight of the queue
+func (q *WeightedQueue) Weight() int64 {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	return q.currentWeight
+}
+
+// Poll retrieves the head of the queue or blocks until an item is available.  The provided exit channel can be closed
+// to interrupt the blocking operation.  Returns the head of the queue and true or nil, false if the poll was
+// interrupted by the closing of the exit channel
+func (q *WeightedQueue) Poll(exit chan struct{}) (WeightedItem, bool) {
+	for {
+		select {
+		case <-exit:
+			return nil, false
+		default:
+
+		}
+
+		q.mu.Lock()
```

Comment:
Curious, why do we need to lock this mutex before checking the length of the queue? (might be worth it a comment)

Reply:
We need to lock here as `q.queue` is unprotected.  I can add a comment to that effect

Reply:
oh, I thought for a second that you were using `WeightedQueue.Len()`